### PR TITLE
conky: Init module

### PR DESCRIPTION
**Pull request by kaleocheng**
### Description

Add new module to enable [conky](https://github.com/brndnmtthws/conky) as service on Linux 

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

File: modules/services/conky.nix
Comment (ambroisie) [inline modules/services/conky.nix:9]: This will just use the `/nix/store/...` path, I think you might want something like `${config.home.homeDirectory}/${config.xdg.configFile."conky/conky.conf".target}` (not sure if `XDG_CONFIG_DIR` is set in a systemd service, otherwise just use that...).

File: modules/services/conky.nix
Comment (ambroisie) [inline modules/services/conky.nix:16]:
```suggestion
      enable = mkEnableOption "Light-weight system monitor";
```

File: modules/services/conky.nix
Comment (ambroisie) [inline modules/services/conky.nix:24]:
```suggestion
          Configuration file path to use instead of
```

**Review by ambroisie — CHANGES_REQUESTED** [review]
I'm not sure I see the point of the `configFile` option, `readFile`/`mkOutOfStoreSymlink` might be all that's needed for this IMO.

File: modules/services/conky.nix
Comment (kaleocheng) [reply modules/services/conky.nix:9]: using the `/nix/store/...` pattern  here as it seems a common way for systemd service in `modules/services` compare to the `config.home.homeDirectory` one. 
- https://github.com/search?q=repo%3Anix-community%2Fhome-manager++path%3Amodules%2Fservices+config.xdg.configFile&type=code
- https://github.com/search?q=repo%3Anix-community%2Fhome-manager++path%3Amodules%2Fservices+config.home.homeDirectory&type=code 

I'm not sure if it's in purpose or just  the legacy reason. let me know if we indeed what to switch to the `home.homeDirectory` one

**Comment by kaleocheng** [comment]
> I'm not sure I see the point of the `configFile` option, `readFile`/`mkOutOfStoreSymlink` might be all that's needed for this IMO.

TBH I'm not using that option, added it just coz `configFile` is kind of a common option.  but you may be right. for the init version it's better to keep it simple. I'm going to remove it .

File: modules/services/conky.nix
Comment (kaleocheng) [reply modules/services/conky.nix:9]: update: use `config.xdg.configHome` seems the good way. I will update it to use that 

-  https://github.com/search?q=repo%3Anix-community%2Fhome-manager++path%3Amodules%2Fservices+config.xdg.configHome&type=code

File: modules/services/conky.nix
Comment (kaleocheng) [inline modules/services/conky.nix:10]: @ambroisie I'm not sure if it is a good idea to get the default config  from command output.  maybe  we just hardcode the default config in here?

Comment (ambroisie) [reply modules/services/conky.nix:10]: Sorry, that's not what I meant when I said one can use `readFile`: you'd introduced an option to give a path to a config file, instead of using the `settings` attribute. Instead one could just do `settings = readFile`.

Definitely a :-1: on doing this, as it's an IFD.

File: modules/lib/maintainers.nix
Comment (ambroisie) [inline modules/lib/maintainers.nix:97]: Please add yourself to the maintainers list in a different commit.

File: modules/services/conky.nix
Comment (ambroisie) [inline modules/services/conky.nix:41]: See first comment:

```suggestion
    xdg.configFile."conky/conky.conf".text = cfg.settings;
```

**Review by ambroisie — CHANGES_REQUESTED** [review]

File: modules/services/conky.nix
Comment (kaleocheng) [reply modules/services/conky.nix:10]: ok, I misunderstood the comments before.   

> Definitely a 👎 on doing this, as it's an IFD.

let me try to  summary  the config/settings part:
- keep  `settings` option only ( as `types.lines` )
- no default config, it means
   - if `settings` is not `""`, then we generate the xdg confile file and append `--config {filepath}` to `ExecStart`
   - else only run `conky`  in  `ExecStart`. in that case Conky will then use the builtin-configs to start. 

so we don't have to run the external command to generate or hardcode a config in here. does it make sense to you?

File: modules/services/conky.nix
Comment (ambroisie) [inline modules/services/conky.nix:11]: The description for `mkEnableOption` is prefixed by `Whether to enable ...`, so something like this would be more appropriate:

```suggestion
      enable = mkEnableOption "conky, a light-weight system monitor";
```

File: modules/services/conky.nix
Comment (ambroisie) [inline modules/services/conky.nix:21]:
```suggestion
          If not set, the default configuration, as described by {command}`conky --print-config`, will be used.
```

**Review by ambroisie — CHANGES_REQUESTED** [review]
Minor nitpicks on the phrasing of the documentation, otherwise LGTM.

**Comment by kaleocheng** [comment]
@ambroisie appreciate for your feedback, I made some updates for the documentation. should I leave it to you next or we need to ping someone else?

**Comment by kaleocheng** [comment]
hmm. I'm considering if  it should be   `program`  module instead of  `service` module. 
asking as I saw the tint2 in `program` https://github.com/nix-community/home-manager/blob/master/modules/programs/tint2.nix

**Comment by kaleocheng** [comment]
@rycee not sure if you are the right person to review this. do you have any suggestion for the next of this PR.

File: modules/services/conky.nix
Comment (rycee) [inline modules/services/conky.nix:15]: I think this is best changed to be `extraConfig` instead of `settings`. Yeah I know the name is a bit odd considering that there are no other options available but that is the convention.

The reason against `settings` is that it is typically referred to _structural settings_ such as a Nix attribute set serialized to JSON. Not raw text lines, is is the case here.

```suggestion
      extraConfig = lib.mkOption {
```

File: modules/services/conky.nix
Comment (rycee) [inline modules/services/conky.nix:48]: Is it necessary to place the configuration file in the user's home directory? Perhaps it would be possible to here do something like
```suggestion
          "--config ${pkgs.writeText "conky.conf" cfg.extraConfig}");
```
instead?

Then there is no need for the file in the home directory _and_ it makes it possible to auto-restart the service when the configuration changes (since the file path changes).

File: tests/modules/services/conky/basic-configuration.nix
Comment (rycee) [inline tests/modules/services/conky/basic-configuration.nix:17]: Can add

```suggestion

      test.stubs.conky = { };
      
```

Then the conky package is replaced by a stub and will not need to be downloaded.

File: modules/services/conky.nix
Comment (rycee) [reply modules/services/conky.nix]: Using `config.xdg.configFile."conky/conky.conf".source` is good here since it will cause the file content to change if the configuration changes. This allows automatic systemd unit restarts.

**Comment by rycee** [comment]
@kaleocheng Apologies for the long delay! I've added a few comments.

File: modules/services/conky.nix
Comment (kaleocheng) [reply modules/services/conky.nix:48]: from tech perspective we don't have to put it in user's home directory. I did that as I saw lots of modules are doing that. e.g. https://github.com/nix-community/home-manager/blob/7e91f2a0ba4b62b88591279d54f741a13e36245b/modules/programs/rofi.nix#L219

but sounds those are kind of "legacy" way, I will follow your comment here to use the writeText instead.

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Pull request merged**
